Null pointer dereference in `U2fAuthenticator::issueSignCommand`

charliewolfe opened the pull request:
#### 9e273ba798318ec4c9710c4ba9b553a91bc0e7bb
<pre>
Null pointer dereference in `U2fAuthenticator::issueSignCommand`
<a href="https://bugs.webkit.org/show_bug.cgi?id=270873">https://bugs.webkit.org/show_bug.cgi?id=270873</a>
<a href="https://rdar.apple.com/124475802">rdar://124475802</a>

Reviewed by Pascoe.

Crash logs indicate it&apos;s possible for `u2fCmd` to be null. Avoid crashing in this case.

* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::issueSignCommand):

Canonical link: <a href="https://commits.webkit.org/276002@main">https://commits.webkit.org/276002@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9721de1833dfe816867f18970318d405c0b31083

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22538 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45917 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46141 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/39635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26340 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19955 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35973 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44080 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16947 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1562 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38849 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47685 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/18536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42757 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19959 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41428 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9677 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/20138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19589 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
```diff
@@ -109,8 +109,11 @@ void U2fAuthenticator::issueSignCommand(size_t index)
     }
     auto u2fCmd = convertToU2fSignCommand(requestData().hash, requestOptions, requestOptions.allowCredentials[index].id, m_isAppId);
+    if (UNLIKELY(!u2fCmd)) {
+        ASSERT_NOT_REACHED();
```
This seems like a somewhat valid state based on `convertToU2fSignCommand`. If a request requires UV and you stick a U2F key in, you'll hit this. 

Let's not ASSERT_NOT_REACHED here and instead log something like
```
U2F_RELEASE_LOG("issueSignCommand: request not convertible to U2F.");
```

and return.

pascoej replied:
I'll filed https://bugs.webkit.org/show_bug.cgi?id=270877 as an enhancement to show a message here as well, but that requires changes to other projects as well. For this PR, I think just the logging is great.

pascoej replied:
also let's do something similar for convertToU2fRegisterCommand (same problem)

pascoej commented on Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
```diff
@@ -84,8 +84,11 @@ void U2fAuthenticator::checkExcludeList(size_t index)
-    U2F_RELEASE_LOG("issueRegisterCommand: Sending %s", base64EncodeToString(*u2fCmd).utf8().data());
-    ASSERT(u2fCmd);
+    if (!u2fCmd) {
+        U2F_RELEASE_LOG("issueSignCommand: request not convertible to U2F.");
```
```suggestion
        U2F_RELEASE_LOG("issueRegisterCommand: request not convertible to U2F.");
```

charliewolfe replied:
whoops

pascoej reviewed (approved):
r+ with one more nit

charliewolfe committed:
Null pointer dereference in `U2fAuthenticator::issueSignCommand`
https://bugs.webkit.org/show_bug.cgi?id=270873
rdar://124475802

Reviewed by Pascoe.

Crash logs indicate it's possible for `u2fCmd` to be null. Avoid crashing in this case.

* Source/WebKit/UIProcess/WebAuthentication/fido/U2fAuthenticator.cpp:
(WebKit::U2fAuthenticator::issueSignCommand):

Canonical link: https://commits.webkit.org/276002@main
